From 39086d914f98b5e77f2c8a72f73e63795d1b0289 Mon Sep 17 00:00:00 2001
From: Harald Hoyer <harald@redhat.com>
Date: Fri, 27 Apr 2012 11:05:30 +0200
Subject: [PATCH] mdraid: fix raid assembly for timeout

array_state was never "|incomplete" but the raid was not active/running
on a degraded array, so the timeout hook was never installed
---
 modules.d/90mdraid/65-md-incremental-imsm.rules | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/modules.d/90mdraid/65-md-incremental-imsm.rules b/modules.d/90mdraid/65-md-incremental-imsm.rules
index 096f34d..1d4d50a 100644
--- a/modules.d/90mdraid/65-md-incremental-imsm.rules
+++ b/modules.d/90mdraid/65-md-incremental-imsm.rules
@@ -47,10 +47,12 @@ GOTO="end_raidstart"
 
 LABEL="do_raidstart"
 
+RUN+="/bin/ln -sf /sbin/md_finished.sh /initqueue-finished/md_finished.sh"
+
 # check if array is not inactive anymore
-TEST=="md/array_state", ATTR{md/array_state}!="|inactive", GOTO="end_raidstart"
+TEST=="md/array_state", ATTR{md/array_state}=="clean", ENV{MD_UUID}=="?*", GOTO="end_raidstart"
 
-RUN+="/bin/sh -c 'ln -s /sbin/md_finished.sh /initqueue-finished/md_finished.sh;/sbin/initqueue --timeout --onetime --name 51-mdraid_start --unique /sbin/mdraid_start'"
+RUN+="/sbin/initqueue --timeout --onetime --name 51-mdraid_start --unique /sbin/mdraid_start"
 
 LABEL="end_raidstart"
 
@@ -68,6 +70,10 @@ GOTO="end_container"
 
 LABEL="do_container"
 
-RUN+="/bin/sh -c 'ln -s /sbin/md_finished.sh /initqueue-finished/md_finished.sh;/sbin/initqueue --timeout --onetime --name 50-mdcontainer_start --unique --name mdcontainer_start-%k /sbin/mdcontainer_start $env{DEVNAME}'"
+RUN+="/bin/sh -c 'ln -s /sbin/md_finished.sh /initqueue-finished/md_finished.sh"
+
+TEST=="md/array_state", ATTR{md/array_state}=="clean", ENV{MD_UUID}=="?*", GOTO="end_container"
+
+RUN+="/sbin/initqueue --timeout --onetime --name 50-mdcontainer_start --unique --name mdcontainer_start-%k /sbin/mdcontainer_start $env{DEVNAME}"
 
 LABEL="end_container"
